src/python/pants/backend/python/tasks/build_local_python_distributions.py

```diff
@@ -135,45 +151,88 @@ def _copy_sources(self, dist_tgt, dist_target_dir):
     all_sources = list(dist_tgt.sources_relative_to_target_base())
     for src_relative_to_target_base in all_sources:
       src_rel_to_results_dir = os.path.join(dist_target_dir, src_relative_to_target_base)
-      safe_mkdir(os.path.dirname(src_rel_to_results_dir))
+      safe_mkdir_for(src_rel_to_results_dir)
       abs_src_path = os.path.join(get_buildroot(),
                                   dist_tgt.address.spec_path,
                                   src_relative_to_target_base)
       shutil.copyfile(abs_src_path, src_rel_to_results_dir)
 
-  def _request_single(self, product, subject):
-    # This is not supposed to be exposed to Tasks yet -- see #4769 to track the
-    # status of exposing v2 products in v1 tasks.
-    return self.context._scheduler.product_request(product, [subject])[0]
-
-  # FIXME(cosmicexplorer): We should be isolating the path to just our provided
-  # toolchain, but this causes errors in Travis because distutils looks for
-  # "x86_64-linux-gnu-gcc" when linking native extensions. We almost definitely
-  # will need to introduce a subclass of UnixCCompiler and expose it to the
-  # setup.py to be able to invoke our toolchain on hosts that already have a
-  # compiler installed. Right now we just put our tools at the end of the PATH.
-  @contextmanager
-  def _setup_py_invocation_environment(self, pythonpath):
-    setup_py_env = self._request_single(
-      SetupPyInvocationEnvironment, self._native_toolchain_instance())
-    env = setup_py_env.as_env_dict()
-    if pythonpath:
+  def _add_artifacts(self, dist_target_dir, shared_libs_product, native_artifact_targets):
+    all_shared_libs = []
+    for tgt in native_artifact_targets:
+      product_mapping = shared_libs_product.get(tgt)
+      base_dir = assert_single_element(product_mapping.keys())
+      shared_lib = assert_single_element(product_mapping[base_dir])
+      all_shared_libs.append(shared_lib)
+
+    for shared_lib in all_shared_libs:
+      basename = os.path.basename(shared_lib.path)
+      # NB: We convert everything to .so here so that the setup.py can just
+      # declare .so to build for either platform.
+      resolved_outname = re.sub(r'\..*\Z', '.so', basename)
```

Comment:
IIRC, `.so` is platform specific. Would this be `.dylib` on OSX? Regardless, make sure OSX coverage via #5920.

Reply:
The python code is assumed to only interact with the shared library through `ctypes`, which does not require the shared library to end in `.dylib` on osx when loading it by file path. As in the comment, this avoids platform-specific code in getting the path to the shared library to declare in the `setup.py` or in the python source.

Reply:
Will add the ctypes integration testing to the osx shard.